File: C++/findRightView.cpp
#include<iostream>
#include<vector>

struct TreeNode{
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
    TreeNode(int x):val(x),left(nullptr),right(nullptr){}
};

class Solution{
    public:
    std::vector<int> solve(std::vector<int>& preOrder, std::vector<int>& inOrder) {
        std::vector<int> res;
        TreeNode* root = reConstructBinaryTree(preOrder, inOrder);
        findRightView(root, res, 0);
        return res;
    }

    void findRightView(TreeNode* root, std::vector<int>& res, int depth) {
        if (root == nullptr) {
            return;
        }
        if (depth >= res.size()) {
            res.push_back(root->val);
        }
        findRightView(root->right, res, depth + 1);
        findRightView(root->left, res, depth + 1);
    }

    TreeNode* reConstructBinaryTree(std::vector<int>& preOrder, std::vector<int>& inOrder) {
        int preStart = 0, preEnd = preOrder.size() - 1;
        int vinStart = 0, vinEnd = inOrder.size() - 1;
        if (preOrder.size() == 0 || inOrder.size() == 0) {
            return nullptr;
        }
        TreeNode* root = new TreeNode(preOrder[preStart]);
        int root_index = 0;
        for (int i = vinStart; i <= vinEnd; i++) {
            if (inOrder[i] == root->val) {
                root_index = i;
                break;
            }
        }
        int left_subtree_size = root_index - vinStart;
        int right_subtree_size = vinEnd - root_index;
        if (left_subtree_size > 0) {
            std::vector<int>preLeft(preOrder.begin() + 1,
                            preOrder.begin() + left_subtree_size + 1);
            std::vector<int>vinLeft(inOrder.begin(), inOrder.begin() + left_subtree_size);
            root->left = reConstructBinaryTree(preLeft, vinLeft);
        } else {
            root->left = nullptr;
        }
        if (right_subtree_size > 0) {
            std::vector<int>preRight(preOrder.begin() + left_subtree_size + 1, preOrder.end());
            std::vector<int>vinRight(inOrder.begin() + left_subtree_size + 1, inOrder.end());
            root->right = reConstructBinaryTree(preRight, vinRight);
        } else {
            root->right = nullptr;
        }
        return root;
    }
};

int main(){
    std::vector<int>preOrder = {1,2,4,5,3};
    std::vector<int>inOrder = {4,2,5,1,3};
    Solution solution;
    TreeNode* reConstructBinaryTree=solution.reConstructBinaryTree(preOrder,inOrder);
    std::vector<int>rightView=solution.solve(preOrder,inOrder);
    std::cout<<"Find right view:";
    for(int val: rightView){
        std::cout<<" "<<val;
    }
    std::cout<<std::endl;
    return 0;
}

File: python/findRightView.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def solve(self , preOrder: list[int], inOrder: list[int]) -> list[int]:
        res=[]
        root =self.reConstructBinaryTree(preOrder,inOrder)
        self.findRightView(root,res,0)
        return res

    def findRightView(self,root,res,depth):
        if not root:
            return None
        if depth>=len(res):
            res.append(root.val)
        self.findRightView(root.right,res,depth+1)
        self.findRightView(root.left,res,depth+1)
        
    def reConstructBinaryTree(self , preOrder: list[int], inOrder: list[int]) -> TreeNode:
        if len(preOrder)==0 or len(inOrder)==0:
            return None
        pre_start=0
        pre_end=len(preOrder)-1
        vin_start=0
        vin_end=len(inOrder)-1
        root=TreeNode(preOrder[pre_start])
        root_index=0
        for i in range(0,vin_end+1):
            if root.val==inOrder[i]:
                root_index=i
                break
        left_subtree_length=root_index-vin_start
        right_subtree_length=vin_end-root_index
        if left_subtree_length>0:
            pre_left=preOrder[1:left_subtree_length+1]
            vin_left=inOrder[0:left_subtree_length]
            root.left=self.reConstructBinaryTree(pre_left,vin_left)
        else:
            root.left=None
        if right_subtree_length>0:
            pre_right=preOrder[left_subtree_length+1:]
            vin_right=inOrder[left_subtree_length+1:]
            root.right=self.reConstructBinaryTree(pre_right,vin_right)
        else:
            root.right=None
        return root
    
preorder=[1,2,4,5,3]
inorder=[4,2,5,1,3]
solution=Solution()
right_view=solution.solve(preorder,inorder)
print(right_view)